### PR TITLE
Feature/full import

### PR DESCRIPTION
**Pull request by julsemaan**
# Description
Create a tool to export and import the PacketFence configuration and database

# Impacts
Its a standalone tool, so not much unless you use it, then it impacts everything!

# Code / PR Dependencies
depends on #6407 

# Delete branch after merge
NO

# Checklist
(REQUIRED) - [yes, no or n/a]
- [WIP] Document the feature
- [N/A] Add unit tests
- [ ] Add acceptance tests (TestLink)

# NEWS file entries
## New Features
* Added a tool to import and export the configuration and database for 10.3.0 and above

### CODE REVIEW

**Comment by julsemaan** [comment]
This is not ready in the sense that it needs more testing and perhaps be merged in devel instead of feature/rhel8 but the code is mostly done unless I missed some stuff to restore

File: addons/full-import/README.md
Comment (extrafu) [inline addons/full-import/README.md]: runned -> run

Comment (julsemaan) [reply addons/full-import/README.md]: I thought it was runnaide

File: addons/full-import/README.md
Comment (extrafu) [inline addons/full-import/README.md]: How about if someone uploaded a jpeg/png file (or something else) to be used on his portal?

File: addons/full-import/README.md
Comment (extrafu) [inline addons/full-import/README.md]: script -> skip

File: addons/full-import/README.md
Comment (extrafu) [inline addons/full-import/README.md]: via scp -> using scp

File: addons/full-import/README.md
Comment (julsemaan) [reply addons/full-import/README.md]: This reminds me I need to grab the file for the `logo` variable of profiles.conf 

But for files referenced in the HTML templates, its pretty hard to track them accurately

I believe that users can use a static/ directory inside their profile-templates/$ID directory which will be migrated through this process.

It may be a bit complex to track assets that are referenced in the HTML template since they won't contain the full path and may be a bit tricky to extract

Comment (extrafu) [reply addons/full-import/README.md]: I think we can limit ourself to the logo in profiles.conf for now

File: addons/full-import/export.sh
Comment (nqb) [inline addons/full-import/export.sh:28]: I suggest to use `export_dir` in place of `build_dir`

**Review by nqb — CHANGES_REQUESTED** [review]
- The file `addons/full-import/find-upgrade-path.pl` doesn't seem to be used anymore.
- We need to be sure packaging is correctly handled for v11 (I can do it and push commits directly to that branch)
- I always used: `set -o nounset -o pipefail -o errexit` at beginning of Bash scripts as a best practice. I think we should at least add `set -o errexit` to be sure any shell scripts will stop in case of an error which is not handle by script itself (like a full filesystem or a permission issue).

**Comment by julsemaan** [comment]
This PR now points on devel

We should plan on testing this tool and then integrating the readme inside our official docs

File: addons/full-import/export.sh
Comment (nqb) [inline addons/full-import/export.sh]: I don't think the `--force` argument is handle in `export.sh`

Comment (julsemaan) [reply addons/full-import/export.sh]: Fixed in 78684d0a6a4e8b0577945bc67529734ff64a8aee

**Comment by julsemaan** [comment]
Rebased against devel, should be all good for testing @nqb

File: addons/full-import/import.sh
Comment (nqb) [inline addons/full-import/import.sh:67]: I think this one can be removed

**Review by nqb — CHANGES_REQUESTED** [review]

File: addons/full-import/import.sh
Comment (nqb) [inline addons/full-import/import.sh:62]: Is it necessary to stop it here because we say in README that we don't support restoring on a cluster ?

File: addons/full-import/database.functions
Comment (nqb) [inline addons/full-import/database.functions]: Indentation looks wrong

File: addons/full-import/database.functions
Comment (nqb) [inline addons/full-import/database.functions]: Indentation looks wrong

File: addons/full-import/database.functions
Comment (nqb) [inline addons/full-import/database.functions]:
```suggestion
    echo "Missing mariabackup command. Please read the documentation of this script in the Installation guide to prepare your environment before running this script"
```

File: addons/full-import/database.functions
Comment (nqb) [inline addons/full-import/database.functions:126]: It looks like `mysql_upgrade` is only done when you did an export on a system which has `mariabackup`. However, standalone installations running 10.3 (without `mariabackup`) need to have their DB upgraded too using this tool.

File: addons/full-import/database.functions
Comment (nqb) [inline addons/full-import/database.functions]:
```suggestion
  db_version=`mysql $mariadb_args $db_name -e "select version from pf_version order by id desc limit 1;" | egrep -o '^[0-9]+\.[0-9]+'`  
```

File: addons/full-import/database.functions
Comment (nqb) [inline addons/full-import/database.functions]: `get_db_upgrade_scripts` ?

File: addons/full-import/configuration.functions
Comment (nqb) [inline addons/full-import/configuration.functions]: `get_config_files` ?

**Review by nqb — CHANGES_REQUESTED** [review]
- import: I suggest to run a `fixpermissions` at end of process:
  * to be compliant with what we do at the moment when restarting PacketFence services
  * to be sure permissions are not different (because we restore files and we change owner)

File: addons/full-import/export.sh
Comment (nqb) [reply addons/full-import/export.sh:37]: I create a new package called `packetfence-export` for v10.3 (Deb and RPM).

This package is a standalone package not dependent to `packetfence` package. It should only be installed on PacketFence installations before 11.0.

This package is built on each commit on devel (can be improve later) and uploaded automatically to Debian and RPM repositories on inverse.ca. A copy to a friendly name like for `packetfence-release` package is done to:
- https://www.packetfence.org/downloads/PacketFence/RHEL8/
- https://www.packetfence.org/downloads/PacketFence/debian/

For 11.0 and next releases, `full-import` directory is part of `packetfence` package.

**Comment by nqb** [comment]
I create a new package called `packetfence-export` for v10.3 (Deb and RPM).

This package is a standalone package not dependent to `packetfence` package. It should only be installed on PacketFence installations before 11.0.

This package is built on each commit on devel (can be improve later) and uploaded automatically to Debian and RPM repositories on inverse.ca. A copy to a friendly name like for `packetfence-release` package is done to:
- https://www.packetfence.org/downloads/PacketFence/RHEL8/
- https://www.packetfence.org/downloads/PacketFence/debian/

For 11.0 and next releases, `full-import` directory is part of `packetfence` package.

File: addons/full-import/export.sh
Comment (nqb) [reply addons/full-import/export.sh]: If you run `export.sh filename` with `filename` a name of a file (without path), it works but I think `filename` is removed at end of process. If somebody used `export.sh --force /path/to/export.tgz`, you got this issue.

Comment (nqb) [reply addons/full-import/export.sh]: Also when using correctly, the `--force` flag doesn't seem to operate as expected because export script continue to copy files from `/root/backup` in place of doing new backups.

Comment (julsemaan) [reply addons/full-import/export.sh]: The `--force` flag actually accepts backups older than 24 hours for cases where your current PF install is broken

In the instructions, we'll state that you must take a backup before running this tool

File: addons/full-import/import.sh
Comment (julsemaan) [reply addons/full-import/import.sh:62]: I'm scared that someone would do it on a cluster they started building and then galera autofix starts doing stuff with the other nodes. Unlikely but its a safety net

File: addons/full-import/database.functions
Comment (julsemaan) [reply addons/full-import/database.functions:126]: When you restore a mysqldump on the server, it already runs the right version of MariaDB and all the tables are recreated so I don't think there is a need to upgrade the MariaDB files since only raw SQL is reimported, not the actual table files

That is unless I'm missing something else mysql_upgrade is doing

File: addons/full-import/export.sh
Comment (nqb) [reply addons/full-import/export.sh]: Sorry I misunderstood usage of `--force`

File: addons/full-import/import.sh
Comment (nqb) [reply addons/full-import/import.sh:62]: I understand, let's keep it.

File: addons/full-import/database.functions
Comment (nqb) [reply addons/full-import/database.functions:126]: Ok, I will check that.

File: addons/full-import/database.functions
Comment (nqb) [inline addons/full-import/database.functions:135]: Could you add a comment to describe what is doing this one-liner ? It's pretty hard to understand and to maintain.

File: addons/full-import/configuration.functions
Comment (nqb) [inline addons/full-import/configuration.functions]: `ipcalc` should be added has a dependency for EL and Debian.

`ipcalc` is packaged in bullseye, seems also ok on EL.

EDIT: it looks `ipcalc` is already present on a ZEN image (CentOS7) but seems to be missing from SPEC.

File: addons/full-import/configuration.functions
Comment (nqb) [inline addons/full-import/configuration.functions]:
```
# ip -br -o a show dev eth0
eth0             UP             10.0.2.15/24 fe80::a00:27ff:fecb:8dc8/64
```
It could simplify to avoid usage of regex.

File: addons/full-import/configuration.functions
Comment (nqb) [inline addons/full-import/configuration.functions]: Idem as above

File: addons/full-import/helpers.functions
Comment (nqb) [inline addons/full-import/helpers.functions]:
```suggestion
    if [ -n "$2" ]; then
```

File: addons/full-import/configuration.functions
Comment (nqb) [inline addons/full-import/configuration.functions]:
```suggestion
        if [ -n "$new_int_choice" ]; then
```

File: addons/full-import/import.sh
Comment (nqb) [inline addons/full-import/import.sh]: Is it on purpose to restore profile templates after upgrade_configuration ?

In the past, a script like `./addons/upgrade/to-8.3-conf-latin1-to-utf8.sh` modify these files.

File: addons/full-import/configuration.functions
Comment (nqb) [inline addons/full-import/configuration.functions]:
```suggestion
function get_configuration_upgrade_scripts() {
```

File: addons/full-import/import.sh
Comment (nqb) [inline addons/full-import/import.sh:132]: Is it necessary to revert `systemctl isolate` ?

**Review by nqb — CHANGES_REQUESTED** [review]

File: addons/full-import/database.functions
Comment (nqb) [reply addons/full-import/database.functions:135]: Same goes for `versions_since` in `helper.functions`

File: addons/full-import/configuration.functions
Comment (nqb) [reply addons/full-import/configuration.functions]: I missed to update calls to this function.

File: addons/full-import/import.sh
Comment (julsemaan) [reply addons/full-import/import.sh:132]: If we do so we start PacketFence and the user may want to do last minute tweaks to his system before starting PF so that's why I'm leaving everything off

File: addons/full-import/import.sh
Comment (julsemaan) [reply addons/full-import/import.sh]: Good point, lets move it

File: addons/full-import/configuration.functions
Comment (julsemaan) [reply addons/full-import/configuration.functions]: ipcalc is part of the base OS, should we still add it to the spec ?

File: addons/full-import/import.sh
Comment (nqb) [reply addons/full-import/import.sh:132]: Ok let's keep as-is.

File: addons/full-import/configuration.functions
Comment (nqb) [reply addons/full-import/configuration.functions]: Necessary on Debian.

**Comment by julsemaan** [comment]
> * import: I suggest to run a `fixpermissions` at end of process:
>   
>   * to be compliant with what we do at the moment when restarting PacketFence services
>   * to be sure permissions are not different (because we restore files and we change owner)

Done

File: addons/full-import/configuration.functions
Comment (julsemaan) [reply addons/full-import/configuration.functions]: Added to both SPEC

**Review by nqb — CHANGES_REQUESTED** [review]
- If I try to restore PF 10.3 export (done on a cluster of EL7) on a 11.0 (Debian 11), I need to install `mariadb-backup` package bundled with PF 10.3 version on Debian 11. However, it conflicts with mariadb-package currently installed with PF 11.0.

EDIT: I propose to use usual [upgrade instructions](https://www.packetfence.org/doc/PacketFence_Upgrade_Guide.html#_database_backup) to take DB backup on 10.3 even if `mariabackup` is better to avoid locks. We can add a note in README to suggest doing export out of business hours. However, export will certainly be done during a maintenance windows so users have been warned about possible disruptions.

I look at installing `mariabackup-10.5` on a 10.3 setup but [MariaDB documentation](https://mariadb.com/kb/en/mariabackup-overview/#compatibility-of-mariabackup-releases-with-mariadb-server-releases) explains that backward compatibility is not always possible.

EDIT2: https://github.com/inverse-inc/packetfence/issues/6092 is another reason to run a manual db_dump using usual upgrade instructions (in place of using backups created by `backup-and-maintenance.sh` script)

**Comment by nqb** [comment]
I prepare a commit to remove requirement on `qpress` because it doesn't seem to be mandatory.

On both backups taken on Deb 9 and EL7 with mariabackup, I don't see any qp files after I used `mbstream`.

On top of that, there is no mention of such requirement in [official Mariabackup documentation.](https://mariadb.com/kb/en/full-backup-and-restore-with-mariabackup/)

File: addons/full-import/import.sh
Comment (nqb) [inline addons/full-import/import.sh]: We should exclude "export.tgz", otherwise, `ls -l` displays `export.tgz` and you think  `export.tgz` is part of the archive.

**Review by nqb — CHANGES_REQUESTED** [review]

File: addons/full-import/configuration.functions
Comment (nqb) [reply addons/full-import/configuration.functions]: There is no option `-m` for `ipcalc` on Debian 11 (ipcalc 0.5, pkg: ipcalc 0.42-2)

Comment (julsemaan) [reply addons/full-import/configuration.functions]: I'll spin up my Debian to figure out a solution

**Comment by nqb** [comment]
In my export, I have three interfaces with corresponding IP addresses
- eth0
- eth1
- eth2

I do my import on a system which has same identifier for interfaces but eth2 is down

=> Import stop without any error.

Caused by return code of: 
```
# ip -br -o a show dev eth2
# echo $?
130
```
Trace:
```
+ echo 'Found interface eth2 in pf.conf'
Found interface eth2 in pf.conf
+ ip a show dev eth2
+ handle_interface_exists eth2
+ interface=eth2
++ get_os_ip_address eth2
++ ip -br -o a show dev eth2
++ head -1
++ egrep -o '[0-9]+\.[0-9]+\.[0-9]+\.[0-9]+'
+ os_ip=

```

**Comment by nqb** [comment]
This is a side effect of `set -o errexit`

**Comment by nqb** [comment]
@fdurand, @JeGoi and @lzammit, any feedback from your tests ?

**Comment by extrafu** [comment]
Documentation looks good.

**Comment by nqb** [comment]
All looks good to me, I'm ready to merge unless someone want to add something.

**Comment by fdurand** [comment]
you have my blessing to merge it

**Pull request merged**
